tailfs/compositefs/compositefs.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// package compositefs provides a webdav.FileSystem that is composi
+package compositefs
+
+import (
+	"io"
+	"log"
+	"os"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/tailscale/xnet/webdav"
+	"tailscale.com/tstime"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/pathutil"
+)
+
+// childrenByName is a slice of *child sorted in name order
+type childrenByName []*Child
+
+func (children childrenByName) Len() int           { return len(children) }
+func (children childrenByName) Swap(i, j int)      { children[i], children[j] = children[j], children[i] }
+func (children childrenByName) Less(i, j int) bool { return children[i].Name < children[j].Name }
+
+// Child is a child filesystem of a CompositeFileSystem
+type Child struct {
+	// Name is the name of the child
+	Name string
+	// FS is the child's FileSystem
+	FS webdav.FileSystem
+	// Available is a function indicating whether or not the child is currently
+	// available.
+	Available func() bool
+}
+
+func (child *Child) defaultAvailable() {
+	if child.Available == nil {
+		child.Available = func() bool { return true }
+	}
+}
+
+// CompositeFileSystem is a webdav.FileSystem that is composed of multiple
+// child webdav.FileSystems. Each child is identified by a name and appears
+// as a folder within the root of the CompositeFileSystem, with the children
+// sorted alphabetically by name.
+//
+// Children in a CompositeFileSystem can only be added or removed via calls to
+// the CompositeFileSystem's SDK methods. From a file system perspective, the
+// root of the CompositeFileSystem acts as read-only, not permitting the
+// addition, removal or renaming of folders.
+//
+// Rename is only supported within a single child. Renaming across children
+// is not supported, as it wouldn't be possible to perform it atomically.
+type CompositeFileSystem interface {
```

Comment:
Is that a standard of ours? I actually like to not export concrete types publicly to give myself the option of changing them later (for example returning different types depending on what the factory method decides). I also find it much easier to read understand the public API by looking at an interface, rather than having to look at all of the methods (with their sometimes long bodies) on a concrete type.